### PR TITLE
fix(rollapp): use ApplyFuncIfNoError in finalize end blocker

### PR DESCRIPTION
**Pull request by danwt**
## Description

In this PR, we protect against errors in endblock
Additionally, we reduce some defensive programming

----

Closes https://github.com/dymensionxyz/dymension/issues/739

**All** items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow-up issues.

PR review checkboxes:

I have...

- [x]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x]  Targeted PR against the correct branch
- [x]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x]  Linked to the GitHub issue with discussion and accepted design
- [x]  Targets only one GitHub issue
- [ ]  Wrote unit and integration tests
- [ ]  Wrote relevant migration scripts if necessary
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters
- [ ]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [ ] Import/Export Genesis
- [ ] Registered Invariants
- [ ] Registered Events
- [ ] Updated openapi.yaml
- [ ] No usage of go `map`
- [ ] No usage of `time.Now()`
- [ ] Used fixed point arithmetic and not float arithmetic
- [ ] Avoid panicking in Begin/End block as much as possible
- [ ] No unexpected math Overflow
- [ ] Used `sendCoin` and not `SendCoins`
- [ ] Out-of-block compute is bounded
- [ ] No serialized ID at the end of store keys
- [ ] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----;

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

### CODE REVIEW

File: x/rollapp/keeper/state_info.go [hunk 1]
```diff
@@ -34,6 +36,17 @@ func (k Keeper) GetStateInfo(
 	return val, true
 }
 
+func (k Keeper) MustGetStateInfo(ctx sdk.Context,
```
Comment (omritoptix) [inline x/rollapp/keeper/state_info.go:39]: why not changing `GetStateInfo` all together to panic? because of the cascading changes required all over?

File: x/rollapp/keeper/block_height_to_finalization_queue.go [hunk 2]
```diff
@@ -7,39 +7,47 @@ import (
...
 			k.SetLatestFinalizedStateIndex(ctx, stateInfoIndex)
 			// call the after-update-state hook
 			keeperHooks := k.GetHooks()
-			err = keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)
+			err := keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)
 			if err != nil {
-				// Failed to call finalization dependent event like ibc packet finalization, invariant breaking. can't proceed
-				return fmt.Errorf("error calling finalized state finalized: rollappID %s, stateInfo: %+v, error %s",
-					stateInfoIndex.RollappId, stateInfo, err.Error())
+				panic(fmt.Errorf("invariant broken: after state finalized must succeed: %w", err))
```
Comment (omritoptix) [inline x/rollapp/keeper/block_height_to_finalization_queue.go:50]: why is this a broken invariant? if the after state finalized doens't succeed, it's not necessarily means an invariant was broken as I see it. in that case it means we failed to finalize e.g rollapp packets so we return an error which will cause revert the changes so far.

File: x/rollapp/keeper/state_info.go (hunk 1 above)
Comment (mtsitrin) [reply x/rollapp/keeper/state_info.go:39]: I think it make sense as u have queries for stateInfo, which u probably don't want to panic for non existing index

File: x/rollapp/keeper/block_height_to_finalization_queue.go [hunk 3]
```diff
@@ -7,39 +7,47 @@ import (
...
+		// we trap at this granularity because we want to avoid iterating inside the
+		// wrapped function (above), and we want the function to be atomic, so
+		// we cannot trap inside finalizePending
+		func(ctx sdk.Context) error {
+			k.finalizePending(ctx, pendingFinalizationQueue)
+			return nil
+		})
+}
+
+func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) {
```
Comment (mtsitrin) [inline x/rollapp/keeper/block_height_to_finalization_queue.go:32]: why u panic in this func instead of returning err?
it's much cleaner. u use cacheCtx on error also, don't have to panic

**Review by mtsitrin — CHANGES_REQUESTED** [review]

File: x/rollapp/keeper/state_info.go (hunk 1 above)
Comment (danwt) [reply x/rollapp/keeper/state_info.go:39]: what @mtsitrin said

File: x/rollapp/keeper/block_height_to_finalization_queue.go (hunk 2 above)
Comment (danwt) [reply x/rollapp/keeper/block_height_to_finalization_queue.go:50]: Fair point. I was putting too much trust in the pre existing comment saying 'invariant breaking'
I will change this one to return an error

File: x/rollapp/keeper/block_height_to_finalization_queue.go (hunk 3 above)
Comment (danwt) [reply x/rollapp/keeper/block_height_to_finalization_queue.go:32]: I have changed the 
`keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)`
to be an error

But I'll keep the panic for the other one

I think it's a good idea to be disciplined about broken invariants, when they break, we should panic, because it means we have a logic bug and can't recover

Does it make sense?

**Review by omritoptix — APPROVED** [review]

**Review by mtsitrin — APPROVED** [review]
